Report more details about encoding and XML errors

vr8hub opened the pull request:
When there's an XML error in a file, lint just dies telling us the filename, but there is no indication what the error might be or where it is. In larger files (stories/chapters/etc.), it can take a while to find the problem. On the Mac, Quicklook will tell you the line/byte of the error, so this hasn't been a huge problem for me, but it would be a lot easier to get more detail direct from lint, and even more so for Windows/Linux.

This adds exceptions to the existing try/except for both encoding errors from the encode, and XML errors from lxml's etree. In a file where I was missing a closing quote on an epub:type, the error would look like this with the change. The "11:227" is the line and byte. (Without the closing quote, it searches the rest of the file for the closing bracket, thinking everything is part of the string. _Technically_ the error is incorrect, but the line/byte is enough information to easily see the problem.)

`Error: Couldn't parse XHTML in file: /Users/vrice/src/books/o-henry_short-fiction/src/epub/text/a-little-local-colour.xhtml, error: <string>:11:227:FATAL:PARSER:ERR_LT_IN_ATTRIBUTE: Unescaped '<' not allowed in attributes values`

acabal commented:
There's a few issues here, some of them caught by pylint:

- Space needed after comma on line 159
- What is `err` on line 159? It's undefined.
- `e` as an exception alias is too short, let's use `ex` to match the rest of the code style
- pylint is also complaining that `XMLSyntaxError` has no member `error_log`
- On line 162 we access an array without checking if the item exists. This should be in another try/catch, or check that the array has > 0 length before continuing.

vr8hub commented:
My apologies on pylint; my other changes have been small enough I haven't run into issues, and I simply forgot to run it. I've fixed everything but the fourth issue (no error_log member).

I don't understand the pylint errors on the object members. In the case of UnicodeError, I'm simply catching a theoretical; I haven't figured out how to generate an encode error to verify it yet. But, the code is straight out of the python documentation (see [here](https://docs.python.org/3.7/library/exceptions.html?highlight=unicodeerror#UnicodeError)). (That's also why I had the err without defining it, doh!)

On the etree error, that is also straight from the lxml docs, and that one I know works, as I've tripped the error.

So, I guess I don't understand what pylint's complaint on those mean.

acabal commented:
Missing a colon on line 164. When you add it pylint has another complaint about style in the if tag. 

UnicodeError may be an abstract class and it might actually throw UnicodeDecodeError. pylint is almost always right so we should aim to fix that error. If you can't actually create the error yourself to test then let's just drop that check until someone actually runs in to that problem.

pylint complains about lxml because lxml is a C extension and pylint can't inspect its source. However I'm not sure why it can't look at a Python exception. Is there an alternate way of accessing the error message such that pylint won't complain? We have to get rid of the complaint.

vr8hub commented:
What the…? I ran pylint. How did that happen?? (Obviously I didn't.)

No problem on UnicodeError. I'll work on trying to get an error to happen, but in the meantime I've removed it.

The code for the lxml error came straight from the lxml documentation. And it works. So I have no idea what pylint is complaining about, or what to do to satisfy its complaint.

Instead, I made a small python script that just called etree and experimented for a while. As it turns out, it looks like the exception message is all we need. I tried several different errors, and got the correct error every time.

I've modified the code accordingly.

There is one pylint error remaining, an Unnecessary "elif" after "continue" on line 979 , but it was already there.

acabal commented:
OK great, thanks!
